plugins/prometheus_plugin/prometheus_plugin.cpp

```diff
@@ -0,0 +1,235 @@
+
+      std::vector<runtime_metric> metrics() {
+        return std::vector{
+          bytes_transferred,
+          num_scrapes
+        };
+      }
+   };
+
+   struct metrics_model {
+      std::vector<std::shared_ptr<Collectable>> _collectables;
+      std::shared_ptr<Registry> _registry;
+      std::vector<std::reference_wrapper<Family<Gauge>>> _gauges;
+      std::vector<std::reference_wrapper<Family<Counter>>> _counters;
+
+      void add_gauge_metric(const runtime_metric& plugin_metric) {
+         auto& gauge_family = BuildGauge()
+               .Name(plugin_metric.family)
+               .Help("")
+               .Register(*_registry);
+         auto& gauge = gauge_family.Add({});
+         gauge.Set(plugin_metric.value);
+
+         _gauges.push_back(gauge_family);
+
+         ilog("Added gauge metric ${f}:${l}", ("f", plugin_metric.family) ("l", plugin_metric.label));
+      }
+
+      void add_counter_metric(const runtime_metric& plugin_metric) {
+         auto &counter_family = BuildCounter()
+               .Name(plugin_metric.family)
+               .Help("")
+               .Register(*_registry);
+         auto &counter = counter_family.Add({});
+         counter.Increment(plugin_metric.value);
+         _counters.push_back(counter_family);
+
+         ilog("Added counter metric ${f}:${l}", ("f", plugin_metric.family) ("l", plugin_metric.label));
+      }
+
+      void add_runtime_metric(const runtime_metric& plugin_metric) {
+         switch(plugin_metric.type) {
+            case metric_type::gauge:
+               add_gauge_metric(plugin_metric);
+               break;
+            case metric_type::counter:
+               add_counter_metric(plugin_metric);
+               break;
+
+            default:
+               break;
+         }
+      }
+
+      void add_runtime_metrics(std::vector<runtime_metric>& metrics){
+         for (auto const& m : metrics) {
+            add_runtime_metric(m);
+         }
+      }
+
+      metrics_model() {
+         _registry = std::make_shared<Registry>();
+         _collectables.push_back(_registry);
+      }
+
+      std::vector<MetricFamily> collect_metrics() {
+         auto collected_metrics = std::vector<MetricFamily>{};
+
+         for (auto&& collectable : _collectables) {
+
+            auto&& metrics = collectable->Collect();
+            collected_metrics.insert(collected_metrics.end(),
+                                     std::make_move_iterator(metrics.begin()),
+                                     std::make_move_iterator(metrics.end()));
+         }
+
+         return collected_metrics;
+      }
+
+      std::string serialize() {
+         const prometheus::TextSerializer serializer;
+         return serializer.Serialize(collect_metrics());
+      }
+   };
+
+   struct prom;
+      struct prometheus_plugin_impl {
+         eosio::chain::named_thread_pool<prom> _prometheus_thread_pool;
```

Comment:
`eosio::chain::named_thread_pool<struct prom> _prometheus_thread_pool;`
Then you don't need the external `struct prom;` above.